src/util/createServerRootMixin.js

```diff
@@ -0,0 +1,255 @@
+   * @param {object} componentInstance the calling component's `this`
+   * @returns {Promise} result of the search, to save for .hydrate
+   */
+  search.findResultsState = function(componentInstance) {
+    let app;
+    return Promise.resolve()
+      .then(() => {
+        const options = {
+          serverPrefetch: undefined,
+          fetch: undefined,
+          _base: undefined,
+          name: 'ais-ssr-root-component',
+        };
+
+        const extended = componentInstance.$vnode
+          ? componentInstance.$vnode.componentOptions.Ctor.extend(options)
+          : Object.assign({}, componentInstance.$options, options);
+
+        app = new Vue(extended);
+
+        app.$options.serverPrefetch = [];
+
+        app.instantsearch.helper = helper;
+        app.instantsearch.mainHelper = helper;
+
+        app.instantsearch.mainIndex.init({
+          instantSearchInstance: app.instantsearch,
+          parent: null,
+          uiState: app.instantsearch._initialUiState,
+        });
+      })
+      .then(() => renderToString(app))
+      .then(() => searchOnlyWithDerivedHelpers(helper))
+      .then(() => {
+        const results = {};
+        walkIndex(app.instantsearch.mainIndex, widget => {
+          results[widget.getIndexId()] = widget.getResults();
+        });
+
+        search.hydrate(results);
+
+        resultsState = Object.keys(results)
+          .map(indexId => {
+            const { _state, _rawResults } = results[indexId];
+            return [
+              indexId,
+              {
+                _state: JSON.parse(JSON.stringify(_state)),
+                _rawResults,
+              },
+            ];
+          })
+          .reduce(
+            (acc, [key, val]) => {
+              acc[key] = val;
+              return acc;
+            },
+            {
+              __identifier: 'stringified',
+            }
+          );
+        return search.getState();
+      });
+  };
+
+  /**
+   * @returns {Promise} result state to serialize and enter into .hydrate
+   */
+  search.getState = function() {
+    if (!resultsState) {
+      throw new Error('You need to wait for findResultsState to finish');
+    }
+    return resultsState;
+  };
+
+  /**
+   * make sure correct data is available in each widget's state.
+   * called in widget mixin with (this.widget, this)
+   *
+   * @param {object} widget The widget instance
+   * @param {object} parent The local parent index
+   * @returns {void}
+   */
+  search.__forceRender = function(widget, parent) {
+    const localHelper = parent.getHelper();
+
+    const results = search.__initialSearchResults[parent.getIndexId()];
+
+    const state = results._state;
+
+    // helper gets created in init, but that means it doesn't get the injected
+    // parameters, because those are from the lastResults
+    localHelper.state = state;
+
+    // TODO: copied from index widget, should be given to all widgets IMO
```

Comment:
> should be given to all widgets IMO

Wdym?

Reply:
this code is written in the index widget, but not passed as an argument, so I need to recreate it here as far as I can tell

Reply:
Yes it makes sense. A comment in the code for "should be given to all widgets IMO" might not be the best place though.